Transport: Compression added

Neofine opened the pull request:
- ConnReader can now decompress frame body on recv()
- Added support for snappy and lz4

Fixes #106

mmatczuk commented on transport/conn.go:
```diff
@@ -327,12 +337,12 @@ func OpenConn(addr string, localAddr *net.TCPAddr, cfg ConnConfig) (*Conn, error
 // WrapConn transforms tcp connection to a working Scylla connection.
 // If error and connection are returned the connection is not valid and must be closed by the caller.
-func WrapConn(conn net.Conn) (*Conn, error) {
+func WrapConn(conn net.Conn, cmpr frame.Compression) (*Conn, error) {
```
Let's pass the whole config here.

mmatczuk commented:
What about requests compression?

mmatczuk commented:
The implementation should use the streaming APIs i.e. io.Reader and io.Writer

Instead of that
```
	// Read body
	if _, err := io.CopyN(c.bufw, c.conn, int64(r.Header.Length)); err != nil {
		r.Err = fmt.Errorf("read body: %w", err)
		return r
	}
	if r.Header.Flags&frame.Compress != 0 {
		if err := c.decompress(&c.buf); err != nil {
			r.Err = fmt.Errorf("decompress body: %w", err)
			return r
		}
	}
```

The `io.CopyN` dst should be the decorating Reader from s2.

Important note we want to avoid allocating those readers for every request and they must be reused.

Same logic applies to writes.

mmatczuk commented on transport/conn.go:
```diff
@@ -488,3 +509,25 @@ func (c *Conn) Close() {
 	return fmt.Sprintf("[addr=%s shard=%d]", c.conn.RemoteAddr(), c.shard)
 }
+
+func decompressSnappy(buf *frame.Buffer) error {
+	from := buf.Bytes()
+	out, err := snappy.Decode(nil, from)
+	if err != nil {
+		return err
+	}
+	buf.Save(out)
+	return nil
+}
+
+func decompressLz4(buf *frame.Buffer) error {
+	sz := uint32(buf.ReadInt())
+	from := buf.Bytes()
+	dst := make([]byte, sz)
+	_, err := lz4.UncompressBlock(from, dst)
+	if err != nil {
+		return err
+	}
+	buf.Save(dst)
+	return nil
+}
```
You can move any compress related funcs to compress.go

mmatczuk commented on transport/conn_integration_test.go:
```diff
@@ -160,3 +160,89 @@ func TestCloseHangingIntegration(t *testing.T) {
+	return &connTestHelper{t: t, conn: conn}
+}
+
+func TestSnappyCompressionIntegration(t *testing.T) {
+	h := newConnCompressionTestHelper(t, frame.Snappy)
+	h.applyFixture()
+
+	defer h.close()
```
It would make more sense to refactor all the test code to a helper function and call that function in a test (named subtest maybe) with compression as a parameter.

mmatczuk reviewed (dismissed):
See the comments inline.

Neofine commented:
Coding progress is little on this branch because it took me quite a long time to figure out that neither s2.Reader nor snappy.Reader works here.

Both of these readers expect to see chunk stream identifier firstly after they start working that means the exact series of bytes is expected when starting communication:

```
 0xff 0x06 0x00 0x00 0x73 0x4e 0x61 0x50 0x70 0x59
```
Which scylla does not send for obvious reasons.

The problems don't end there, Readers seem to expect first byte as chunk identifier (which tells what data to expect from chunk) and 3 consecutive more to be length of our data.

From my testing first byte in body is the length of data chunk, so our specifications don't match (I haven't checked bodies of greater length than single byte value).

The only solution to this problem I see is to make our own Reader (I haven't gone in-depth with Writers yet), which
uses snappy.Decode in its body, I think this wouldn't be far from original s2.Reader and snappy.Reader but
it would have less features and constraints.

What do you think?

mmatczuk commented:
Dear @Neofine please note the `ReaderOption` that you add to new reader.
Namely `ReaderIgnoreStreamIdentifier`, we also want `ReaderAllocBlock` to limit the decompression buffer size.
Let me know if that helps.

Neofine commented:
Compression couldn't be done with neither s2 nor lz4 Readers/Writers because of their funky signature
i.e. s2.Reader needs also crc32 hash of uncompressed data so we can't make our own translation writer.

In compress.go I have added writers and readers to snappy and lz4. 
I am not very satisfied with the result as readers can't be chained with other readers, now we have do:

```go
for {
	if _, err := io.CopyN(c.bufw, c.dec, int64(r.Header.Length)); err != nil {
		if err.Error() != "END" {
			r.Err = fmt.Errorf("read body: %w", err)
			return r
		} else {
			break
		}
	}
}
```
to copy uncompressed data into c.bufw. And because of reader restrictions with my knowledge it can't be done much prettier (we can return n = -1 instead of err = "END" but for loop needs to be there).

Writers seem easy in nature and chainable.

Both snappy and lz4 work on simple test cases. To be sure of its correctness we could add some bigger tests.

Neofine commented:
Compression couldn't be done with neither s2 nor lz4 Readers/Writers because of their funky signature
i.e. s2.Reader needs also crc32 hash of uncompressed data so we can't make our own translation writer.

In compress.go I have added writers and readers to snappy and lz4. 
Readers not to loop need to return error, so I have added dummy ErrEnd and when performing io.Copy we get this:

```go
if _, err := io.Copy(c.bufw, c.dec); err != nil {
	if !errors.Is(err, ErrEnd) {
		r.Err = fmt.Errorf("read body: %w", err)
		return r
	}
}
```
Any suggestions to make it prettier are appreciated.

Writers seem easy in nature and chainable.

Both snappy and lz4 work on simple test cases. To be sure of its correctness we could add some bigger tests.

Neofine commented on transport/compress.go:
```diff
@@ -0,0 +1,303 @@
+		}
+
+		// TODO: Why we need to do this?
+		if zeroes(r.buf[:r.l]) {
```
This is a band-aid solution.
Scylla when sending `Ready` response seem to also send body of 5 zeroes.
I am not sure of its cause.

mmatczuk commented on frame/buffer.go:
```diff
@@ -6,8 +6,11 @@ import (
 // CopyBuffer writes data to w until the buffer is drained or an error occurs.
-func CopyBuffer(b *Buffer, w io.Writer) (n int64, err error) {
-	return b.buf.WriteTo(w)
+func CopyBuffer(b *Buffer, w io.Writer) (n int, err error) {
```
Why is this changed?

mmatczuk replied:
Why is this changed?

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,232 @@
+var (
+	ErrComprCorrupt     = errors.New("compression: corrupt input")
+	ErrComprUnsupported = errors.New("compression: unsupported input")
+	errComprEnd         = errors.New("end")
```
Can we use nil instead of it?

mmatczuk commented:
Some nice improvements @Neofine.

I pushed a refactoring changes - I wanted to merge this but there is a problem.

The issue is that we have here a growing buffer. I'd like to suggest that following:

- Forget the "babushka" style wrapping of readers and writers
- Add a helper type that would expose `Copy(dst Writer, src Reader) (written int64, err error)`
- The constructor parameters are compress or decompress, type of compression, compression buffer size
- The size of buffer should be in ConnConfig, I suggest default of 64Kb but please double check that
- There must be `buf` and `tmp` `[]byte` fields of fixed size initiated on construction
- The copy would follow [Copy with buffer code from io pkg](https://github.com/golang/go/blob/78fb1d03d39e8357e4790a9f0788ef0a8e7d8ae1/src/io/io.go#L404)
    - Fill the buffer (buf)
    - Call a function: compress or decompress (best done with function pointer)
    - Write tmp
- On the perf side you can detect if reader is `bytes.Buffer` and access the bytes directly no need to copy them again

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,232 @@
+	binary.BigEndian.PutUint32(l, uint32(len(p)))
+
+	enc = p[:9]
+	enc = append(enc, l...)
```
Would be better to append shifted bytes directly, we do that in other places already.

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,172 @@
+func newCompression(compress, snappy bool, bufSize int) *compr {
+	c := &compr{
+		buf: make([]byte, bufSize),
+		tmp: make([]byte, bufSize),
```
This should probably be larger / smaller depending on compression type.
Before you calculated the size the temporary buffer.

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,172 @@
+		buf: make([]byte, bufSize),
+		tmp: make([]byte, bufSize),
+	}
+	switch compress {
```
Do not use a switch over bool.

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,172 @@
+
+func (c *compr) Copy(dst io.Writer, src io.Reader) (written int64, err error) {
+	var (
+		toRead int
```
how about size?

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,172 @@
+	tmp    []byte
+	header []byte
+
+	compressor func(int) (int, error)
```
I'd call that process or something since it can compress or decompress.

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,172 @@
+		}
+	}
+
+	if len(c.header) != 0 {
```
It would be easier to read if c.header != nil

mmatczuk commented on transport/compress.go:
```diff
@@ -0,0 +1,172 @@
+	if len(c.header) != 0 {
+		nrh, erh = dst.Write(c.header)
+		if erh != nil {
+			if ok {
```
What's that?

Neofine commented:
I may be mistaken but from my observations scylla doesn't seem to divide data into the actual chunks, even when asked to do so (with `'chunk_length_kb': 64`), after making (not compressed) insertion of 600k characters - approx 0.6 MB I query compressed data and first bytes of chunks from both lz4 and snappy are telling that length of uncompressed data of nearest chunk is 600k.

In both gocql and rust drivers we have respectively:
```go
func (s SnappyCompressor) Decode(data []byte) ([]byte, error) {
	return snappy.Decode(nil, data)
}
```
```rust
...
Compression::Snappy => snap::raw::Decoder::new()
            .decompress_vec(comp_body)
            .map_err(|_| FrameError::FrameDecompression),
...
```

where both functions work on full body of a frame. In golang `snappy.Decode(nil, data)` where `data` consists of two concatenated chunks results in error so gocql surely expects a single big chunk.

s2.Reader has option of reading data chunk by chunk:

```go
chunkLen := int(r.buf[1]) | int(r.buf[2])<<8 | int(r.buf[3])<<16
```

but we resigned from using it as scylla doesn't give us data compatible with it.

With my current understanding of the situation we can't avoid sudden allocations of slices of (in this case) 600k items,  what should we do now?

Kulezi commented on transport/conn.go:
```diff
@@ -113,18 +114,25 @@ func (c *connWriter) send(r request) error {
-	if _, err := frame.CopyBuffer(&c.buf, c.conn); err != nil {
-		return err
+	var err error
+	if r.Compress {
```
c.compr being nil probably should return some meaningful error instead of nil

Kulezi commented on transport/conn.go:
```diff
@@ -432,18 +460,27 @@ func validateKeyspace(keyspace string) error {
+const cqlVersion = "3.0.0"
 
-func (c *Conn) init() error {
+func (c *Conn) init(cfg ConnConfig) error {
```
Why give cfg to a *Conn method when Conn owns it already?

Kulezi commented on transport/compress.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func (c *compr) decodeLz4(read int) (int, error) {
+	buf := c.buf.Bytes()
+	if read < 4 {
+		return 0, ErrComprCorrupt
+	}
```
check should be done before declaring buf

Kulezi commented on transport/compress.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func (c *compr) Compress(dst io.Writer, src *bytes.Buffer) (written int64, err error) {
+	c.buf = *src
```
should c.buf be overwritten like that?

Kulezi replied:
n/m

Michal-Leszczynski commented on transport/compress_test.go:
```diff
@@ -0,0 +1,158 @@
+	return out
+}
+
+func TestCompression(t *testing.T) {
```
It seems like random tests / fuzzing could be an interesting addition for testing compression.

Choraden commented on transport/compress.go:
```diff
@@ -0,0 +1,146 @@
+		return int64(nrh), erh
+	}
+
+	nr, er := dst.Write(c.tmp.Bytes()[:n])
```
How about we wrap this line with some WriteCompressed to hide tmp variable, which is used in codec function. 
What I mean is that codec function silently uses tmp, but it is not documented. I would opt for hiding it completely.

Kulezi replied:
Added documentation about tmp usage, I believe now it's ok

Choraden commented on transport/compress.go:
```diff
@@ -0,0 +1,146 @@
+		return 0, err
+	}
+
+	nr, er := dst.Write(c.tmp.Bytes()[:n])
```
Same here.

Choraden reviewed (commented):
left some comments

Kulezi commented on transport/compress.go:
```diff
@@ -0,0 +1,146 @@
+	return int64(nr), er
+}
+
+func newCompression(compress bool, algo frame.Compression, bufSize int) (*compr, error) {
```
Changed to newCompr to fit the struct name + added comment about the bool.
